src/Microsoft.Azure.WebJobs.Extensions.ServiceBus/Listeners/ServiceBusScaleMonitor.cs

```diff
@@ -0,0 +1,337 @@
+            {
+                bool queueTimeIncreasing =
+                    IsTrueForLastN(
+                        metrics,
+                        NumberOfSamplesToConsider,
+                        (prev, next) => prev.EventQueueTime <= next.EventQueueTime);
+                if (queueTimeIncreasing)
+                {
+                    status.Vote = ScaleVote.ScaleOut;
+                    _logger.LogInformation($"Queue time is increasing for '{_entityPath}'.");
+                    return status;
+                }
+            }
+
+            bool queueLengthDecreasing =
+                IsTrueForLastN(
+                    metrics,
+                    NumberOfSamplesToConsider,
+                    (prev, next) => prev.TotalNewEventCount > next.TotalNewEventCount);
+            if (queueLengthDecreasing)
+            {
+                status.Vote = ScaleVote.ScaleIn;
+                _logger.LogInformation($"Queue length is decreasing for '{_entityPath}'.");
+                return status;
+            }
+
+            bool queueTimeDecreasing = IsTrueForLastN(
+                metrics,
+                NumberOfSamplesToConsider,
+                (prev, next) => prev.EventQueueTime > next.EventQueueTime);
+            if (queueTimeDecreasing)
+            {
+                status.Vote = ScaleVote.ScaleIn;
+                _logger.LogInformation($"Queue time is decreasing for '{_entityPath}'.");
+                return status;
+            }
+
+            _logger.LogInformation($"Service Bus entity '{_entityPath}' is steady.");
+
+            return status;
+        }
+
+        private bool TimeToLogWarning()
+        {
+            DateTime currentTime = DateTime.UtcNow;
+            bool timeToLog = currentTime >= _nextWarningTime;
+            if (timeToLog)
+            {
+                _nextWarningTime = currentTime.AddHours(1);
+            }
+            return timeToLog;
+        }
+
+        private static bool IsTrueForLastN(IList<ServiceBusTriggerMetrics> samples, int count, Func<ServiceBusTriggerMetrics, ServiceBusTriggerMetrics, bool> predicate)
+        {
+            // Walks through the list from left to right starting at len(samples) - count.
+            for (int i = samples.Count - count; i < samples.Count - 1; i++)
+            {
+                if (!predicate(samples[i], samples[i + 1]))
+                {
+                    return false;
+                }
+            }
+
+            return true;
+        }
+    }
+
+    internal class ServiceBusMetricsData
+    {
+        public ServiceBusMetricsData(Message message, long activeMessageCount, long deadLetterCount, int partitionCount)
+        {
+            Message = message;
+            ActiveMessageCount = activeMessageCount;
+            DeadLetterCount = deadLetterCount;
+            PartitionCount = partitionCount;
+        }
+
+        public Message Message { get; set; }

```

Comment:
This isn't part of the serializable metrics contract right? I.e. shouldn't be serialized to metrics table storage? If not, please remove, and instead pass the Message into your create metrics helper directly.